db/active_replicator_checkpointer.go

```diff
@@ -187,29 +166,199 @@ func calculateCheckpointSeq(expectedSeqs []string, processedSeqs map[string]stru
 	return seq
 }
 
-func (c *Checkpointer) setCheckpoint(seq string) {
+const checkpointDocIDPrefix = "checkpoint/"
+
+// fetchCheckpoints tries to fetch since values for the given replication by requesting a checkpoint on the local and remote.
+// If we fail to fetch checkpoints, we remove any existing checkpoints, and start from zero.
+// If we fetch mismatched checkpoints, we'll pick the lower of the two, and remove the higher checkpoint.
+func (c *Checkpointer) fetchCheckpoints() error {
+	base.TracefCtx(c.ctx, base.KeyReplicate, "fetchCheckpoints()")
+
+	localSeq, localRev, err := c.getLocalCheckpoint()
+	if err != nil {
+		return err
+	}
+	base.DebugfCtx(c.ctx, base.KeyReplicate, "got local checkpoint: %q %q", localSeq, localRev)
+	c.lastLocalCheckpointRevID = localRev
+
+	remoteSeq, remoteRev, err := c.getRemoteCheckpoint()
+	if err != nil {
+		return err
+	}
+	base.DebugfCtx(c.ctx, base.KeyReplicate, "got remote checkpoint: %q %q", remoteSeq, remoteRev)
+	c.lastRemoteCheckpointRevID = remoteRev
+
+	// If localSeq and remoteSeq match, we'll use this value.
+	checkpointSeq := localSeq
+
+	// Determine the lowest sequence if they don't match
+	if localSeq != remoteSeq {
+		base.DebugfCtx(c.ctx, base.KeyReplicate, "sequences mismatched, finding lowest of %q %q", localSeq, remoteSeq)
+		localSeqVal, err := parseIntegerSequenceID(localSeq)
+		if err != nil {
+			return err
+		}
+
+		remoteSeqVal, err := parseIntegerSequenceID(remoteSeq)
+		if err != nil {
+			return err
+		}
+
+		if remoteSeqVal.Before(localSeqVal) {
+			err := c.removeLocalCheckpoint(localRev)
```

Comment:
On a mismatch, do we actually want to remove the local checkpoint here (and the remote below)?  If the replication (or SG) were to crash after removing the checkpoint, I think that means that on restart the replication would be rolled back to zero.  Should we be updating to the lower value instead?

Reply:
Ah yeah, fair comment. I'll fix that later.

Reply:
@adamcfraser Updated to address this comment.

Still pending adding rollback tests, but up to you whether to merge as-is now, or after those are committed tomorrow.